### ISSUE TITLE
B018: false positive

### ISSUE THREAD

**Issue by spaceone** [astral-sh/ruff#5006]
```
try:
    unicode
except NameError:
    unicode = str
```
is not a useless expression `B018 Found useless expression. Either assign it to a variable or remove it.`

**Comment by dhruvmanila** [comment]
Hmm, interesting usage. So, it's checking if there's a `unicode` name in the scope and assigning `str` to it if there's none. Until we find a solution for this, you could use the `globals()` dictionary:

```python
unicode = globals().get('unicode', str)
```

**Comment by charliermarsh** [comment]
I think the most we would do here is tailor the rule message to this specific case (e.g., suggest an alternative pattern when scoped with a `NameError` handler). Can you share a realistic example where this code is necessary, and can't be refactored to something that doesn't rely on an unbound name?

**Comment by spaceone** [comment]
the above code is the real code we are using to support Python2.
other examples in our code:
```
        with self.assertRaises(AttributeError):
            obj.props.unknown  # noqa: B018

        with pytest.raises(AttributeError):
            obj.props.unknown  # noqa: B018

   def validate(self, value):
        # type: (str) -> object
        o = urlsplit(value)
        o.port  # may raise ValueError  # noqa: B018
        return o.scheme in {"http", "https"}
```

**Comment by dhruvmanila** [comment]
Related: https://github.com/astral-sh/ruff/issues/3831

**Comment by charliermarsh** [comment]
My recommendation is still to use `_ = obj.props.unknown` to indicate that the useless expression is intentional.

**Comment by charliermarsh** [comment]
If it's explicitly to support Python 2, can you not use a `sys.version_info` check instead of using the error-based control flow?

**Comment by charliermarsh** [comment]
I'm comfortable with the behavior we have for the originating example. Whether we want to special-case attribute accesses in some way will be resolved via https://github.com/astral-sh/ruff/issues/3831.

**Comment by DanielNoord** [comment]
@charliermarsh We have code at work that looks like this:

```python
A: list


def func():
    global A
    try:
        A
    except NameError:
        print("Help")
        A = []


func()
func()
```

We use the global in jobs to append to a global list, but we are never sure which worker (which execute the jobs) will start up first and therefore which worker will call the setup function `func()`.
Would you reconsider excluding raising this message if it is specifically guarded by a `NameError`? Doesn't that explicit guard show that the writer is aware of the chance of an exception being raised and thus no bug is introduced?

**Comment by charliermarsh** [comment]
@DanielNoord - I think the main question for me is whether there's a way to (reasonably) restructure the code to avoid the "useless expression". For example, would something like this work?

```python
A: list

def func():
    global A
    if 'A' not in globals():
        print("Help")
        A = []

func()
func()
```

(I know this is a greatly simplified example -- I'm mostly trying to decide if we need to support this pattern, or if we can instead suggest a reasonable alternative.)

**Comment by DanielNoord** [comment]
> @DanielNoord - I think the main question for me is whether there's a way to (reasonably) restructure the code to avoid the "useless expression". For example, would something like this work?

 I can see how another lint would be to never use `globals()` or `global` for that matter. I think the pattern is quite bad anyway :smile: But having that global available is quite helpful as well and I think sometimes a little ugly code for a lot of usability doesn't hurt.

Personally I think the `try ... except` shows in a very concise and explicit way that you know you're introducing a possibility for `NameError` but are also handling it. Thus, it doesn't hurt maintainability of the code too much and (as far as I can see) doesn't introduce an actual bug into your code. I can add a `# noqa: B018` but I believe that introduces more uncertainty for the reader as they now need to lookup what B018 is.

**Comment by DanielNoord** [comment]
@charliermarsh Sorry for tagging again, but have you given this a thought? And if so, what is your verdict about re-opening this issue?
